### PR TITLE
Accept Proc in the `:on` argument

### PR DESCRIPTION
**Pull request by rafaelsales**

### CODE REVIEW

File: lib/retriable.rb [hunk 1]
```diff
@@ -73,4 +69,23 @@ def retriable(opts = {})
       end
     end
   end
+
+  def matched_exception?(on, exception)
```
[the remark this answers is not in the document]
Comment (apurvis) [reply lib/retriable.rb]: should be `private`

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,7 @@
 ## HEAD
 
 * Fancier README
+* Add ability for `:on` argument to accept a `Hash` where the keys are exception types and the values are a single or array of `Proc` that evaluates whether to retry exception.
```
Comment (apurvis) [inline CHANGELOG.md]: would be a 3.2.0 release probable

File: README.md [hunk 3]
```diff
@@ -95,9 +95,10 @@ Here are the available options, in some vague order of relevance to most common
 - An `Exception` class (retry every exception of this type, including subclasses)
 - An `Array` of `Exception` classes (retry any exception of one of these types, including subclasses)
 - A `Hash` where the keys are `Exception` classes and the values are one of:
-  - `nil`(retry every exception of the key's type, including subclasses)
+  - `nil` (retry every exception of the key's type, including subclasses)
+  - A single `Proc` (retries exceptions ONLY if it returns truthy)
   - A single `Regexp` pattern (retries exceptions ONLY if their `message` matches the pattern)
-  - An array of patterns (retries exceptions ONLY if their `message` matches at least one of the patterns)
+  - An array of Proc and/or Regexp (retries exceptions ONLY if at least one exception matches the Proc or Regexp)
```
Comment (apurvis) [inline README.md]: there might be a collision with https://github.com/kamui/retriable/pull/47

File: lib/retriable.rb [hunk 4]
```diff
@@ -73,4 +69,23 @@ def retriable(opts = {})
...
+    return true unless on.is_a?(Hash)
+
+    on.any? do |expected_exception, matchers|
+      next false unless exception.is_a?(expected_exception)
+      next true if matchers.nil?
+
+      Array(matchers).any? do |matcher|
+        if matcher.is_a?(Regexp)
+          exception.message =~ matcher
+        elsif matcher.respond_to?(:call)
```
Comment (apurvis) [inline lib/retriable.rb]: why not `is_a?(Proc)`?

File: CHANGELOG.md (hunk 2 above)
Comment (rafaelsales) [reply CHANGELOG.md]: I thought `HEAD` would be changed to master on the changelog only when the owner would release it. Should I put it there now?

Comment (kamui) [reply CHANGELOG.md]: Yeah, I like the changelog stuff in PRs to go into HEAD. It allows me to decide how many commits to group into a release and also lets me decide whether it's a major/minor/patch level version change.

**Comment by kamui** [comment]
Hmm, can you get the specs to pass?

File: lib/retriable.rb [hunk 5]
```diff
@@ -73,4 +69,26 @@ def retriable(opts = {})
       end
     end
   end
+
+  module Internal
+    def self.matched_exception?(on, exception)
```
Comment (rafaelsales) [inline lib/retriable.rb]: If I just make this method private on the Retriable module, we can't call it... This private Internal module was the only workaround I came up with... other ideas?

**Comment by rafaelsales** [comment]
@kamui done

File: lib/retriable.rb (hunk 5 above)
Comment (kamui) [reply lib/retriable.rb]: Try `private_class_method :matched_exception?`.

Comment (rafaelsales) [reply lib/retriable.rb]: that was it!

File: CHANGELOG.md (hunk 2 above)
Comment (apurvis) [reply CHANGELOG.md]: yeah i've actually adopted the `HEAD` approach myself at this point, so this was more a way to propose a version number when stuff is released eventually than any actual complaint

Comment (apurvis) [reply CHANGELOG.md]: trivial but "are a single or array" is not quite english.  "and the values are either `Proc`s or arrays of `Proc`s that will evaluate to `true` when the exception should be retried"

File: README.md [hunk 6]
```diff
@@ -129,12 +130,13 @@ Retriable.retriable on: [Timeout::Error, Errno::ECONNRESET] do
 end
 ```
 
-You can also specify a Hash of exceptions where the values are a list or single Regexp pattern.
+You can also specify a Hash of exceptions where the values are a list or single Proc or Regexp pattern.
```
Comment (apurvis) [inline README.md]: How about:
> You can also specify a Hash where the keys are `Exception` subclasses and the values are either:
> 1. A `Regexp` (or array of `Regexp`s).  If any of the `Regexp`s match the exception's message, the block will be retried.
> 1. A `Proc` (or array of `Proc`s) that evaluates the exception being handled and returns `true` if the block should be retried.  If any of the procs in the list return `true`, the block will be retried.
> 
> You can also mix and match `Proc`s and `Regexp`s in an Array

File: lib/retriable.rb [hunk 7]
```diff
@@ -73,4 +69,24 @@ def retriable(opts = {})
...
+
+    on.any? do |expected_exception, matchers|
+      next false unless exception.is_a?(expected_exception)
+      next true if matchers.nil?
+
+      Array(matchers).any? do |matcher|
+        if matcher.is_a?(Regexp)
+          exception.message =~ matcher
+        elsif matcher.is_a?(Proc)
+          matcher.call(exception)
```
Comment (apurvis) [inline lib/retriable.rb]: would it be worth passing any additional params to the `Proc`?  Maybe like [the `on_retry.call()`](https://github.com/kamui/retriable/blob/7b9da542e89d9e2ffaf4f1803fc34128a2f806e7/lib/retriable.rb#L70)

File: spec/retriable_spec.rb [hunk 8]
```diff
@@ -313,6 +313,34 @@ class SecondTestError < TestError; end
       expect(exceptions[3].class).must_equal StandardError
     end
 
+    it '#retriable retries with a hash exception where the value is a proc ' \
+       'that returns true' do
```
Comment (apurvis) [inline spec/retriable_spec.rb]: not sure why this part isn't just on the previous line? there's plenty of room even given the style guide limitations

File: spec/retriable_spec.rb [hunk 9]
```diff
@@ -313,6 +313,34 @@ class SecondTestError < TestError; end
...
+          tries += 1
+          raise TestError, 'something went wrong'
+        end
+      end.must_raise TestError
+
+      expect(tries).must_equal 2
+    end
+
+    it '#retriable does not retry with a hash exception where the value is a ' \
+       'proc that returns false' do
```
Comment (apurvis) [inline spec/retriable_spec.rb]: same as above - just put this in the previous string?

**Comment by apurvis** [comment]
@rafaelsales this PR has merge conflicts now

**Pull request closed without merging**